fluss-server/src/main/java/com/alibaba/fluss/server/coordinator/MetadataManager.java

```diff
@@ -385,78 +393,97 @@ public void createPartition(
+                                        tablePath),
+                                e);
+                    }
 
-        try {
-            long partitionId = zookeeperClient.getPartitionIdAndIncrement();
-            // register partition assignments to zk first
-            zookeeperClient.registerPartitionAssignment(partitionId, partitionAssignment);
-            // then register the partition metadata to zk
-            zookeeperClient.registerPartition(tablePath, tableId, partitionName, partitionId);
-            LOG.info(
-                    "Register partition {} to zookeeper for table [{}].", partitionName, tablePath);
-        } catch (KeeperException.NodeExistsException nodeExistsException) {
-            if (!ignoreIfExists) {
-                throw new PartitionAlreadyExistsException(
-                        String.format(
-                                "Partition '%s' already exists for table %s",
-                                partition.getPartitionQualifiedName(), tablePath));
-            }
-        } catch (Exception e) {
-            throw new FlussRuntimeException(
-                    String.format(
-                            "Register partition to zookeeper failed to create partition %s for table [%s]",
-                            partitionName, tablePath),
-                    e);
+                    try {
+                        int bucketCount = partitionAssignment.getBucketAssignments().size();
+                        // currently, every partition has the same bucket count
+                        int totalBuckets = bucketCount * (partitionNumber + 1);
+                        if (totalBuckets > maxBucketNum) {
+                            throw new TooManyBucketsException(
+                                    String.format(
+                                            "Adding partition '%s' would result in %d total buckets for table %s, exceeding the maximum of %d buckets.",
+                                            partition.getPartitionName(),
+                                            totalBuckets,
+                                            tablePath,
+                                            maxBucketNum));
+                        }
+                    } catch (TooManyBucketsException e) {
+                        throw e;
+                    } catch (Exception e) {
+                        throw new FlussRuntimeException(
+                                String.format(
+                                        "Failed to check total bucket count for table %s",
+                                        tablePath),
+                                e);
+                    }
+
+                    try {
+                        long partitionId = zookeeperClient.getPartitionIdAndIncrement();
+                        // register partition assignments to zk first
+                        zookeeperClient.registerPartitionAssignment(
```

Comment:
Can we wrap `registerPartitionAssignment` and `registerPartition` into a zk transation? We let zk to do the concurrent controll so that we won't need the lock. We will have clean code. 
The only bad is that we may fire mutiple zk transation for same partition. But it shouldn't happen frequently, so I think it's fine since the current implemenation also will fire mutiple zk operations for same partition.